keps/1397-evictions-in-background/README.md

```diff
@@ -0,0 +1,1147 @@
+- Implement a multi-level sorting: Allow plugins to first sort pods based
+  on eviction-in-progress quality and then based on configured sorting plugins.
+  Preferring pods that are already in a process of eviction over other pods
+  is a crucial step for reducing unnecessary disruptions.
+- Internal counters aware of eviction in progress: At the beginning of each
+  descheduling loop all internal counters will take into account pods
+  that are either annotated with `descheduler.alpha.kubernetes.io/eviction-in-progress`
+  or are present in the cache of to-be-evicted pods.
+
+#### v1alphaN or beta
+- Replace the current eviction API with the new
+  [evacuation API](https://github.com/kubernetes/enhancements/pull/4565) by default.
+  Both `descheduler.alpha.kubernetes.io/request-evict-only` and
+  `descheduler.alpha.kubernetes.io/eviction-in-progress` will be deprecated.
+  Existing deployments will need to remove `descheduler.alpha.kubernetes.io/request-evict-only`
+  annotation to utilize the new evacuation API.
+  Otherwise, the eviction API will be used as a fallback.
+  Workload owners utilizing the new evacuation API are expected to replace both annotations
+  with [evacuation.coordination.k8s.io/priority_${EVACUATOR_CLASS}: ${PRIORITY}/${ROLE}](https://github.com/kubernetes/enhancements/blob/2d7dbab6c95575d68395c30fa789f87d9a2c3c01/keps/sig-apps/4563-evacuation-api/README.md#evacuator).
+  Important: the new annotation have different semantics than the deprecated ones.
+  Users are strongly recommended to review the evacuation API proposal for necessary details.
+- The evacuation API expects the evacuation request to be reconciled during each descheduling cycle.
+  Given each cycle starts from scratch the descheduler needs to first list
+  all evacuation requests and sort the pods accordingly in each plugin.
+  Other components may create an evacuation request as well.
+  The descheduler needs to take this into account.
+  E.g. by creating a snapshot (per cycle, per strategy) to avoid inconsistencies.
+  An evacuation request may exist without a corresponding pod
+  before the evacuation controller garbage collects the request.
+- The descheduler needs to keep a cache of evacuation API requests (CRs)
+  and remove itself from finalizers if the request is no longer needed.
+  E.g. `evacuation.coordination.k8s.io/instigator_descheduler.sigs.k8s.io` as a finalizer.
+  The removal is allowed only when the evacuation request has its cancellation policy set `Allow`.
+- Read the eviction status from the [evacuation CR status](https://github.com/kubernetes/enhancements/pull/4565).
+
+#### Code changes
+
+The code responsible for invoking the eviction API is located
+under `sigs.k8s.io/descheduler/pkg/descheduler/evictions` under private `evictPod` function.
+The function can be generalized into a generic interface:
+
+```go
+type PodEvacuator interface {
+    EvacutePod(ctx context.Context, pod *v1.Pod) error
```

Comment:
what guarantee we have each plugin will call this interface? Isn't it better to override/change the Evictor interface?

Reply:
The Evictor interface will get replaced by `PodEvacuator` interface under the hood. Currently, all plugins are expected to invoke `d.handle.Evictor().Evict` method. The method will invoke the evacuation API once available and get renamed to a more suitable name.